python/sglang/srt/layers/sampler.py

```diff
@@ -121,6 +126,33 @@ def forward(
 
         return batch_next_token_ids
 
+    def _apply_custom_logit_processor(
+        self, logits: torch.Tensor, sampling_batch_info: SamplingBatchInfo
+    ):
+        """Apply custom logit processors to the logits.
+        This function will modify the logits in-place."""
+
+        for (
+            processor_str,
+            batch_mask,
+        ) in sampling_batch_info.custom_logit_processor.items():
+            # Get the processor from the string representation
+            processor = CustomLogitProcessor.from_str(processor_str)
```

Comment:
This should be done once per request instead of per-token. You can cache it and use an integer as index later.